src/type/validate.ts

```diff
@@ -144,6 +144,44 @@ function validateRootTypes(context: SchemaValidationContext): void {
         (subscriptionType as any).astNode,
     );
   }
+
+  if (queryType && mutationType && queryType.name === mutationType.name) {
+    context.reportError(
+      'All root types must be different, ' +
+        queryType.name +
+        ' is already used for "query" and cannot also be used for "mutation".',
+      getOperationTypeNode(schema, OperationTypeNode.MUTATION) ??
+        mutationType.astNode,
+    );
+  }
+
+  if (
+    queryType &&
+    subscriptionType &&
+    queryType.name === subscriptionType.name
```

Comment:
While I also prefer the name equal check, it seems like a `queryType  === subscriptionType` might also be sufficient?

Reply:
I wasn't sure about if `graphql-js` always keeps the same reference to the type objects, and I thought checking the name of the type is safer as it also prevents issues when doing stuff like this for whatever reason:
```js
const query = new GraphQLObjectType(...);
const anotherQuery = new GraphQLObjectType(query.toConfig());
expect(query === anotherQuery).toBeFalse();
```

Reply:
There is a separate check for different objects of the same type name, so we can/should have this check be of object identity imo

Reply:
Good point @yaacovCR, you can't construct a `GraphQLSchema` object when two types have the same name. I changed it to check for referential equality here.